omegaconf/omegaconf.py

```diff
@@ -399,7 +398,7 @@ def resolver_wrapper(
                     f"`register_new_resolver()` instead (see "
                     f"https://github.com/omry/omegaconf/issues/426 for migration instructions)."
                 )
-            key = args
+            key = args_str
```

Comment:
is this sensitive to whitespace and other insignificant changes:
```
c1 : ${cached:a, b}
c2 : ${cached:a,b}
```
I think those will have the same key. If that's true:
I think a way around it is to use string representation of the parsed input as key:
```
key = "||".join(str(x) for x in args)
```

Reply:
Not sure I understand. This is *not* sensitive to whitespaces like the one you showed (in fact, I intentionally introduced an extra whitespace in the `cached_2` doc example to show it implicitly).

It will however be sensitive to other kinds of whitespaces. For instance `${foo}` and `${ foo }` both evaluate to the same value but have different string literal representations.

We can't use `args` unless we do another 180 to base the key on the arguments and not the string literals (for instance, for an interpolation we could have `10` in `args` and `"${foo}"` in `args_str`)

Reply:
Oh, okay: args_str is a Tuple of strings. I thought it was a single string.
Why do we even want to support `${  foo  }` as an alternative to `${foo}` ?

We can either be more restrictive or maybe cannonize the passed input (which sounds a bit hard to me).

Reply:
> Why do we even want to support `${ foo }` as an alternative to `${foo}` ?

I think that back when we were discussing the grammar you mentioned ignoring whitespaces around braces. I can imagine situations where it could make things a bit more readable, e.g. `${ ${foo:bar} }`.
That being said, no objection if you prefer to remove this ability (it doesn't seem like it was allowed in 2.0).

> We can either be more restrictive or maybe cannonize the passed input (which sounds a bit hard to me).

I wouldn't canonize it (too hard indeed). The example I gave is just one case, there are plenty more (ex: `1000` vs `1_000`, `"foo"` vs `foo`, `{a: b}` vs `{a:b}`, etc.)
I don't think we can ever hope the string literal representation to be a unique identifier of the value that is being passed to the resolver. If you prefer, we can switch to the string representation of the input value as you suggested (though I would use `repr()` instead of `str()`). It's closer to what I had before, but without the added complexity of handling different data types.
Let me know what seems best to you.

Reply:
> I think that back when we were discussing the grammar you mentioned ignoring whitespaces around braces. I can imagine situations where it could make things a bit more readable, e.g. ${ ${foo:bar} }.
That being said, no objection if you prefer to remove this ability (it doesn't seem like it was allowed in 2.0).

I think if we can find a way to make it work in a more unique way we can keep this flexibility.


> I wouldn't canonize it (too hard indeed). The example I gave is just one case, there are plenty more (ex: 1000 vs 1_000, "foo" vs foo, {a: b} vs {a:b}, etc.)
I don't think we can ever hope the string literal representation to be a unique identifier of the value that is being passed to the resolver. 

Okay, let's drop this idea.

> If you prefer, we can switch to the string representation of the input value as you suggested (though I would use repr() instead of str()). It's closer to what I had before, but without the added complexity of handling different data types.
Let me know what seems best to you.

I still want this to be static, so it should not resolve interpolations (including custom resolvers).
Maybe something like this:
```
1 -> 1 - > "1"
1_0  -> 10 - > "10"
${foo} -> ${foo} -> ${foo}
${foo:1,2} -> ${foo, [1,2]} -> "${foo, [1,2]}"
${foo:1,"2 "} -> ${foo, [1,"2 "]} ->  "${foo, ['1','2 ']}"
```
But I suspect this would actually require converting the parse tree to an AST as an intermediate step.
Going through an AST would make things easier to test as well, but it's a pretty big change.

Reply:
Do we have an intermediate representation of an interpolation as something with the data of:
```
${foo} : Interpolation(key="foo")
${bar:a,b,c} : CustomResolver(name="bar", args=["a","b","c"]) 
${bar:${z:1,2}} : CustomResolver(name="bar", args=[CustomResolver(name="z", args=[1,2])]) 
```
?

Reply:
> Do we have an intermediate representation of an interpolation as something with the data of:
> (...)

At the moment we don't keep track of these.

I don't think it would be too hard to achieve what you have in mind, but I expect it to be a bit annoying (mostly due to quoted strings -- also should make sure to only do it for resolvers with cache enabled to avoid adding overhead everywhere).

It's something I'd rather not get into for 2.1, considering that IMO it adds little benefit over the current basic approach based on raw string literals (it's only going to be useful in niche cases).

Reply:
This is more similar to what I did in Hydra.
The visitor is returning an AST, which is easier to deal with:
1. It's easier to test ([example](https://github.com/facebookresearch/hydra/blob/master/tests/test_overrides_parser.py#L209-L234)).
2. It's easier to reason about.
3. It allows canonization

In Hydra, primitive rules are returning primitive results, except quoted strings  which returns QuotedString object, which allows reconstruction of the original input if needed. ([link](https://github.com/facebookresearch/hydra/blob/master/tests/test_overrides_parser.py#L495-L643)).

Number 2 is important:
For example, what would this return?
```
x: 'foo ${bar}'
```

I hope that would be quoted string `foo ${bar}` that is later parsed (like any other string) for interpolations in it, something like `QuotedString(quote="'", text="foo ${bar}")`.

Note that this will allow the eliminations of the callbacks, disentangling the tests of the parsing from the test of interpolation behavior.

Reply:
> The visitor is returning an AST

The way I see it, the tree that the visitor is going through is already a kind of AST, isn't it? If we return yet another tree then we'll need to visit it anyway at some point in order to resolve it and get the actual value we care about.
(not saying there couldn't be benefits in "transforming" the tree provided by ANTLR into a different one that would be more convenient for our needs though)

> Number 2 is important:
> For example, what would this return?
> 
> ```
> x: 'foo ${bar}'
> ```

If this is an excerpt from a YAML file, then this is something like a `ConfigValue([TopLevelStr("foo "), NodeInterpolation("bar")])`

If this is a key/value pair while parsing a dict, it would be something like `dictKeyValuePair(dictKey=ID("x"), element=QuotedString(quote="'", ...)`. I can see two options for `...`:
- Either it's something like you wrote: `text="foo ${bar}"`
- Or we actually parse it: `content=ConfigValue([TopLevelStr("foo "), NodeInterpolation("${bar}")])`

The advantage of the second option, from a cache perspective, is that it would make it more robust to irrelevant changes in the quoted content (ex: using `${ bar }` instead of `${bar}`).

This is why I mentioned quoted strings as making things a bit annoying. Also, ideally I'd want `${foo:bar}` and `${foo:"bar"}` to use the same cache key (but that shouldn't be too hard considering that unquoted strings are quite limited in what they support).

> Note that this will allow the eliminations of the callbacks, disentangling the tests of the parsing from the test of interpolation behavior.

Right now we could easily test that the parsing runs without error separately from any callback, if we wanted to.
It's true that we don't have tests for the correctness of the parsing though. Maybe I'm lazy but given the number of tests, I find it easier to provide the final expected value than the expected parse tree ;)

> I think the overhead is limited to interpolations in general. so we are already protected from significant slowdown.

Right, may not be a meaningful concern.

Bottom line: I agree that we could build our own tree to replace the one from ANTLR. I'd rather not do that now myself though.